### PR TITLE
Issue 13 resolve overlap between add methods

### PR DESCRIPTION
**Pull request by yt-ms**
* [X] fix https://github.com/Midnighter/structurizr-python/issues/13
* [X] description of feature/fix
* [X] tests added/passed
* [X] add an entry to the [next release](../CHANGELOG.rst)

- Model is now much more passive in its add methods, with most of the work being done in the parent element (e.g. Container).
- Replaced add() with += for more convenient syntax.
- Small change in the way hydrate() works, in that the Container registers itself with the model rather than SoftwareSystem doing that for it.  This is so that Container has the model when we start to hydrate Components into it.

-------------------

THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE Apache License v.2.0. YOU MAY OBTAIN A COPY OF THE LICENSE AT https://www.apache.org/licenses/LICENSE-2.0.

THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: src/structurizr/model/container.py [hunk 1]
```diff
@@ -70,7 +71,7 @@ class Container(StaticStructureElement):
         technology: The technology associated with this container
                     (e.g. Apache Tomcat).
         tags: A comma separated list of tags associated with this container.
-        components: The set of components within this container.
+        components: The set of components within this container.  Do not modify directly.
```
Comment (yt-ms) [inline src/structurizr/model/container.py]: How concerned are we about people modifying components directly?  We could provide a read-only property (returning a read-only set or iterable) instead, but is it worth it?

File: src/structurizr/model/container.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 """Provide a container model."""
 
+from __future__ import annotations
```
Comment (yt-ms) [inline src/structurizr/model/container.py]: This doesn't work in Python 3.6.  Before I work around, are we still planning to support 3.6?

Comment (yt-ms) [reply src/structurizr/model/container.py]: Actually, I've taken the incompatible stuff out for the moment so I don't clutter the PRs.

File: src/structurizr/model/model.py [hunk 3]
```diff
@@ -209,38 +209,29 @@ def add_software_system(self, software_system=None, **kwargs) -> SoftwareSystem:
         self.software_systems.add(software_system)
         return software_system
 
-    def add_container(
-        self, container: Optional[Container] = None, **kwargs
-    ) -> Container:
+    def add_container(self, container: Container) -> Container:
```
Comment (yt-ms) [inline src/structurizr/model/model.py]: I guess there's a question now as to whether there's any point having `Model.add_container()` and `Model.add_component()` or whether it'd be better for `Container` and `Component` just to call `Model._add_element()` directly.

File: tests/unit/model/test_container.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+import pytest
+
+from structurizr.model import Component, Container, Model
+
+
+_model = (
+    Model()
+)  # Have to create outside the fixture so it doesn't get garbage-collected.
```
Comment (yt-ms) [inline tests/unit/model/test_container.py]: I'm going to look at whether this hack can be avoided by instead mocking the SoftwareSystem of the Container.

File: tests/unit/model/test_software_system.py [hunk 5]
```diff
@@ -18,10 +18,23 @@
...
 import pytest
 
+from structurizr.model.container import Container
 from structurizr.model.model import Model
 from structurizr.model.software_system import SoftwareSystem
 
 
+_model = (
+    Model()
+)  # Have to create outside the fixture so it doesn't get garbage-collected.
```
Comment (yt-ms) [inline tests/unit/model/test_software_system.py]: I'll look at whether I can mock to avoid this hack.

File: src/structurizr/model/container.py (hunk 1 above)
Comment (Midnighter) [reply src/structurizr/model/container.py]: As much as it's possible in Python I would like to disallow modifying components (and other container attributes) directly indeed. It is extra effort yes but it has the well known benefit that we can modify the underlying container type (list, set, dict) to our liking whenever we please.

Comment (yt-ms) [reply src/structurizr/model/container.py]: Agreed.  This will be quite a big change across many classes if we want to enforce it consistently (especially as it's not just about making the property read-only, but also that to the outside world it looks like a read-only collection).

Another example is in BigBank.py you can see adding to the tags collection directly.

Comment (yt-ms) [reply src/structurizr/model/container.py]: I've just refactored this specific case (Container.components) to try out one approach.  See what you think.

File: src/structurizr/model/container.py [hunk 6]
```diff
@@ -99,41 +98,68 @@ def __init__(
...
         self.technology = technology
-        self.components = set(components)
+        self._components = set(components)
 
         self.tags.add(Tags.CONTAINER)
 
+    @property
+    def components(self) -> Iterable[Component]:
+        """Return read-only list of child components."""
+        return [c for c in self._components]
```
Comment (yt-ms) [inline src/structurizr/model/container.py]: Trying out an approach to making key properties read-only to the outside world.  See what you think.

Went for list rather than generator as it's easier to use outside and they'll never be huge.

File: src/structurizr/model/component.py [hunk 7]
```diff
@@ -82,7 +82,7 @@ class Component(StaticStructureElement):
     def __init__(
         self,
         *,
-        parent: "Container",
+        parent: "Container" = None,
```
Comment (Midnighter) [inline src/structurizr/model/component.py]: If you set a default of `None` please annotate the type as optional.

```suggestion
        parent: Optional["Container"] = None,
```

File: src/structurizr/model/container.py [hunk 8]
```diff
@@ -80,10 +79,10 @@ class Container(StaticStructureElement):
     def __init__(
         self,
         *,
-        parent: "SoftwareSystem",
+        parent: "SoftwareSystem" = None,
```
Comment (Midnighter) [inline src/structurizr/model/container.py]:
```suggestion
        parent: Optional["SoftwareSystem"] = None,
```

File: src/structurizr/model/container.py (hunk 6 above)
Comment (Midnighter) [reply src/structurizr/model/container.py]: Right now I'm undecided what's more useful, a set or a list? So either

```py
return self._components.copy()
```

or simply

```py
return list(self._components)
```

File: src/structurizr/model/container.py [hunk 9]
```diff
@@ -99,41 +98,68 @@ def __init__(
...
+        """Add a newly constructed component to this container."""
+        # TODO: once we move past python 3.6 change to proper return type via
+        # __future__.annotations
+        if component in self.components:
+            # Nothing to do
+            return self
+
+        if self.get_component_with_name(component.name):
+            raise ValueError(
+                f"Component with name {component.name} already exists for {self}."
```
Comment (Midnighter) [inline src/structurizr/model/container.py]: I'm not a native speaker but possibly 'already exists on/in' rather than 'for'?

File: src/structurizr/model/model.py (hunk 3 above)
Comment (Midnighter) [reply src/structurizr/model/model.py]: I have my head in a lot of other projects right now, so I don't have the clearest vision. If there is no special logic that we need in `add_container`, etc. and we don't foresee any special logic being added in the near future, then I think refactoring to a single `Model.add_element` would be fine.

Having `Model.add_element` would also allow us to simplify the `__iadd__` and `add_*` methods in the other modules.

Comment (Midnighter) [reply src/structurizr/model/model.py]: I do think we should be consistent with either the `__iadd__` method calling the `add_*` methods or the other way around. (See `SoftwareSystem` below.) We could add a 'protected' helper method(s) called by both.

File: tests/unit/model/test_software_system.py [hunk 10]
```diff
@@ -18,10 +18,32 @@
...
+
+    def add_container(self, container):
+        """Simulate the model assigning IDs to new elements."""
+        if not container.id:
+            container.id = "id"
+        container.set_model(self)
+        pass
+
+
+_model = MockModel()
```
Comment (Midnighter) [inline tests/unit/model/test_software_system.py]: I didn't fully understand your earlier comment about this being garbage collected. The way I'd do it is:

```py
@pytest.fixture(scope="function")
def model():
   return MockModel()

@pytest.fixture(scope="function")
def model_with_system(model):
    software_system = SoftwareSystem(name="Sys")
    software_system.set_model(model)
    return model


def test_software_system_add_constructed_container(model_with_system):
    system = model.get_system...
    ...
```

Plus the docstrings :wink:

And possibly a separate fixture like you currently have but without a model set on it for some of the tests.

Comment (yt-ms) [reply tests/unit/model/test_software_system.py]: What I found is that unit tests failed unexpectedly because the Element only holds a weak ref to the model so by the time we needed it then the model had been collected.  I think your suggestion would have the same problem.

File: src/structurizr/model/container.py (hunk 6 above)
Comment (yt-ms) [reply src/structurizr/model/container.py]: As a consumer, and as this is now read-only, I think list works better so I'll go with that, thanks.

File: tests/unit/model/test_software_system.py (hunk 10 above)
Comment (Midnighter) [reply tests/unit/model/test_software_system.py]: Here I am suggesting to pass the model down to the test function and the model holds the system. So it should not be garbage collected.

File: src/structurizr/model/model.py (hunk 3 above)
Comment (yt-ms) [reply src/structurizr/model/model.py]: Have a look at what I've done to Model in my latest commit and let me know if you think this approach is good.

File: tests/unit/model/test_software_system.py (hunk 10 above)
Comment (yt-ms) [reply tests/unit/model/test_software_system.py]: Ah, sorry - I'd not spotted that the test methods were now taking the model rather than the system.  I'll have a look at that.  Thanks.

Comment (yt-ms) [reply tests/unit/model/test_software_system.py]: OK, I've given that a go.  See what you reckon.

File: src/structurizr/model/software_system.py [hunk 11]
```diff
@@ -74,19 +74,41 @@ def __init__(self, *, location: Location = Location.Unspecified, **kwargs) -> No
...
+            )
+
+        if container.parent is None:
+            container.parent = self
+        elif container.parent is not self:
+            raise ValueError(
+                f"Container with name {container.name} already has parent {container.parent}. Cannot add to {self}."
+            )
+        self.containers.add(container)
+        self.get_model().add_container(container)
```
Comment (yt-ms) [inline src/structurizr/model/software_system.py]: This shouldn't work.  I'll fix it but also look at the tests.

Comment (yt-ms) [reply src/structurizr/model/software_system.py]: Fixed now.

File: src/structurizr/model/container.py [hunk 12]
```diff
@@ -99,41 +98,68 @@ def __init__(
...
+        if component.parent is None:
+            component.parent = self
+        elif component.parent is not self:
+            raise ValueError(
+                f"Component with name {component.name} already has parent "
+                f"{component.parent}. Cannot add to {self}."
+            )
+        self._components.add(component)
+        model = self.get_model()
+        model += component
```
Comment (Midnighter) [inline src/structurizr/model/container.py]: I'd shorten this to:
```suggestion
        self.get_model()  += component
```
btw, there is also a property, so if you prefer you can write
```suggestion
        self.model += component
```

File: src/structurizr/model/model.py [hunk 13]
```diff
@@ -106,20 +105,29 @@ def __init__(
...
         self._id_generator = SequentialIntegerIDGenerator()
 
     def __contains__(self, element: Element):
+        """Return True if the element is in the model."""
         return element in self.get_elements()
 
+    @property
+    def software_systems(self) -> Set[SoftwareSystem]:
+        """Return the software systems in the model."""
+        return set([e for e in self.get_elements() if isinstance(e, SoftwareSystem)])
```
Comment (Midnighter) [inline src/structurizr/model/model.py]: I'd suggest using a set comprehension:

```suggestion
        return {e for e in self.get_elements() if isinstance(e, SoftwareSystem)}
```

File: src/structurizr/model/model.py [hunk 14]
```diff
@@ -106,20 +105,29 @@ def __init__(
...
+
+    @property
+    def people(self) -> Set[Person]:
+        """Return the people in the model."""
+        return set([e for e in self.get_elements() if isinstance(e, Person)])
```
Comment (Midnighter) [inline src/structurizr/model/model.py]:
```suggestion
        return {e for e in self.get_elements() if isinstance(e, Person)}
```

File: src/structurizr/model/software_system.py [hunk 15]
```diff
@@ -74,19 +74,43 @@ def __init__(self, *, location: Location = Location.Unspecified, **kwargs) -> No
...
+            name=name, description=description, technology=technology, **kwargs
         )
+        self += container
+        return container
+
+    def __iadd__(self, container: Container) -> "SoftwareSystem":
+        """Add a newly constructed container to this system and register with its model."""
+        # TODO: once we move past python 3.6 change to proper return type via __future__.annotations
+        if container in self.containers:
+            # Nothing to do
```
Comment (Midnighter) [inline src/structurizr/model/software_system.py]:
```suggestion
```

File: src/structurizr/model/software_system.py [hunk 16]
```diff
@@ -74,19 +74,43 @@ def __init__(self, *, location: Location = Location.Unspecified, **kwargs) -> No
...
+        if container.parent is None:
+            container.parent = self
+        elif container.parent is not self:
+            raise ValueError(
+                f"Container with name {container.name} already has parent "
+                f"{container.parent}. Cannot add to {self}."
+            )
+        self.containers.add(container)
+        model = self.get_model()
+        model += container
```
Comment (Midnighter) [inline src/structurizr/model/software_system.py]:
```suggestion
        self.model += container
```

File: src/structurizr/model/software_system.py [hunk 17]
```diff
@@ -74,19 +74,43 @@ def __init__(self, *, location: Location = Location.Unspecified, **kwargs) -> No
...
+        return self
+
+    def get_container_with_name(self, name: str) -> Container:
+        """Return the container with the given name, or None."""
+        return next((c for c in self.containers if c.name == name), None)
```
Comment (Midnighter) [inline src/structurizr/model/software_system.py]: I like this implementation. It hints at a possible change in future when we make `self._containers` a dict keyed by name.

```py
return self.containers.get(name)
```

**Review by Midnighter — APPROVED** [review]
I'm very happy with this implementation now. Some very minor comments that you can implement at your discretion.

File: src/structurizr/model/model.py (hunk 13 above)
Comment (yt-ms) [reply src/structurizr/model/model.py]: I didn't even know you could do that!  Thanks.

File: src/structurizr/model/container.py (hunk 12 above)
Comment (yt-ms) [reply src/structurizr/model/container.py]: Unfortunately that doesn't work as self.model is read-only, which is irritating.  I've added an `add()` method to Model instead which makes this code less ugly, but I'm not sure whether it'd confuse.

**Comment by yt-ms** [comment]
If you're good with the latest changes then I think this is ready for merge and I'll pick up something else.

File: src/structurizr/model/container.py (hunk 12 above)
Comment (Midnighter) [reply src/structurizr/model/container.py]: Hmm, that is curious. Sorry I didn't expect that. I must admit that I don't like the addition of the `add` method that you made in https://github.com/Midnighter/structurizr-python/pull/33/commits/880affe39482ef893a8540ce5fbb51054cbd2bcd I'd rather that you revert those changes and then I'll merge.

Comment (yt-ms) [reply src/structurizr/model/container.py]: Yep, agreed.  Done.

**Comment by Midnighter** [comment]
:tada: An important step forwards. Cheers :slightly_smiling_face:
